File: scala/src/Combinations.scala

import scala.annotation._

/**
 * Form all possible combinations of n mutually distinguishable
 * elements m at a time where each combination is a selection of m
 * elements from the n without regard to their order of arrangement.
 */
object Combinations {
  /** Test driver.
    * Usage: scala Combinations A B C D E
    */
  def main( elements: Array[String] ) {
    // Assume the elements are arranged from most to least
    // prominent. The elements are lexical in this example but
    // there is no restriction that they be lexical.

    // n is the number of elements
    val n = elements.length

    // t is the possible number of non-repeating combinations: 2**n
    val t = BigInt( 1 ) << n

    // human order for all m
    println( "human order, m=0 to m=n:" )
    for ( m <- elements.indices )
      humanOrderM( elements, m, binomialCoefficient( n, m ), t - 1 )
    println( "" )

    // machine order for all m
    println( "machine order, m=0 to m=n:" )
    machineOrder( elements, t )
    println( "" )


    val m = 2

    // human order for a specific m
    println( "human order, m=" + m + " for example:" )
    humanOrderM( elements, m, binomialCoefficient( n, m ), t - 1 )
    println( "" )

    // machine order for a specific m
    println( "machine order, m=" + m + " for example:" )
    machineOrderM( elements, m, binomialCoefficient( n, m ) )
  }


  /** Human order for a specific m. Start with bits = 2**n - 1. Recurse
    * through all combinations of non-repeating elements, decrementing
    * by one.
    * @param elements The set of elements
    * @param m The number of elements to be combined
    * @param u The possible number of non-repeating combinations for m
    * @param bits The bit representation of the current non-repeating combination
    * @param count The current count of combinations with m elements
    */
  @tailrec
  private def humanOrderM( elements: Array[String], m: Int, u: BigInt, bits: BigInt, count: BigInt = 0 ) {
    //  check whether count is less than the possible number of non-repeating combinations for m
    if ( count < u ) {
      // Check whether this is a combination of m elements.
      if ( bits.bitCount == m ) {
        // Store and list the current combination. Reverse the order
        // to take advantage of the fact that less significant bits
        // change faster when incrementing or decrementing.
        listCombination( storeCombination( elements.reverse, bits ).reverse )

        humanOrderM( elements, m, u, bits - 1, count + 1 )
      }
      else {
        humanOrderM( elements, m, u, bits - 1, count )
      }
    }
  }
  

  /** Machine order for a specific m. Start with bits = 0. Recurse
    * through all combinations of non-repeating elements, incrementing
    * by one.
    * @param elements The set of elements
    * @param t The possible number of non-repeating combinations
    * @param bits The bit representation of the current non-repeating combination
    */
  @tailrec
  private def machineOrder( elements: Array[String], t: BigInt, bits: BigInt = 0 ) {
    if ( bits < t ) {
      // Store and list the current combination.
      listCombination( storeCombination( elements, bits ) )

      machineOrder( elements, t, bits + 1 )
    }
  }


  /** Machine order for a specific m. Start with bits = 0. Recurse
    * through all combinations of non-repeating elements, incrementing
    * by one.
    * @param elements The set of elements
    * @param m The number of elements to be combined
    * @param u The possible number of non-repeating combinations for a specific m
    * @param bits The bit representation of the current non-repeating combination
    * @param count The current count of combinations with m elements
    */
  @tailrec
  private def machineOrderM( elements: Array[String], m: Int, u: BigInt, bits: BigInt = 0, count: BigInt = 0 ) {
    //  check whether count is less than the possible number of non-repeating combinations for m
    if ( count < u ) {
      // Check whether this is a combination of m elements.
      if ( bits.bitCount == m ) {
        // Store and list the current combination.
        listCombination( storeCombination( elements, bits ) )

        machineOrderM( elements, m, u, bits + 1, count + 1 )
      }
      else {
        machineOrderM( elements, m, u, bits + 1, count )
      }
    }
  }


  /** Binomial coefficient for optimization with large n. Uses factorials.
    * @param n The total number of elements
    * @param m The number of elements to be combined
    */
  private def binomialCoefficient( n: Int, m: Int ): BigInt =
    factorial( n ) / ( factorial( m ) * factorial( n - m ) )


  /** Factorial used for binomial coefficient
    * @param n The number for which the factorial is calculated
    * @param accumulator The current factorial
    */
  @tailrec
  private def factorial( n: Int, accumulator: BigInt = 1 ): BigInt =
    if ( n == 0 ) accumulator else factorial( n - 1, accumulator * n )



  /** Store the combination
    * @param elements The set of elements
    * @param bits The bit representation of the current non-repeating combination
    */
  private def storeCombination( elements: Array[String], bits: BigInt = 0 ): Array[String] = {
    val combination = new Array[String]( elements.length )
    val empty = "-"

    // For all n elements test whether element j is present.
    for ( j <- elements.indices ) {
      // If present add it to the current combination.
      if ( bits.testBit( j ) )
        combination( j ) = elements( j )
      else
        combination( j ) = empty
    }
    combination
  }


  /** Process the given combination. Here we just write it to the console.
    *
    * @param combination The combination to process
    */
  private def listCombination( combination: Array[String] ) {
    val n = combination.length

    // For this combination list the elements that are present.
    for ( j <- 0 until n )
    {
      print( combination.apply(j).toString )
      print( " " ) // element separator
    }
    println( "" ) // end of this combination
  }
}

File: java/src/Combinations.java
import java.lang.reflect.Array;
import java.math.BigInteger;
import java.util.ArrayList;


/**
 * Form all possible combinations of n mutually distinguishable
 * elements m at a time where each combination is a selection of m
 * elements from the n without regard to their order of arrangement.
 */
public class Combinations
{
    /** Test driver.
     * Usage: java -jar Combinations.jar a b c d e
     */
    public static void main( String[] elements )
    {
        // Assume the elements are arranged from most to least
        // prominent. The elements are lexical in this example but
        // there is no restriction that they be lexical.
        
        // n is the number of elements
        int n = Array.getLength( elements );

        // t is total number of nonrepeating combinations: 2^n
        BigInteger t = BigInteger.ONE.shiftLeft( n );

        
        // human order for all m
        System.out.println( "human order, m=0 to m=n:" );
        for ( int m = 0; m <= n; ++m )
            humanOrder( elements, n, t, m );
        System.out.println( "" );

        // machine order for all m
        System.out.println( "machine order, m=0 to m=n:" );
        machineOrder( elements, n, t );
        System.out.println( "" );

        
        int m = 2;
        
        // human order for a specific m
        System.out.println( String.format( "human order, m=%d for example:", m ) );
        humanOrder( elements, n, t, m );
        System.out.println( "" );

        // machine order for a specific m
        System.out.println( String.format( "machine order, m=%d for example:", m ) );
        machineOrder( elements, n, t, m );
    }    

    
    /** human order
     * @param elements The set of elements
     * @param n The total number of elements
     * @param t The total number of nonrepeating combinations
     * @param m The number of elements to be combined
     */
    private static void humanOrder( Object[] elements, int n, BigInteger t, int m )
    {
        // total number of combinations with m elements
        BigInteger b = binomialCoefficient( n, m );        
        // current count of combinations with m elements
        BigInteger q = BigInteger.ZERO;
            
        // Loop over all 2^n combinations, decrementing by one.
        for ( BigInteger i = t.subtract( BigInteger.ONE );
              i.compareTo( BigInteger.ZERO ) >= 0 && q.compareTo( b ) < 0;
              i = i.subtract( BigInteger.ONE ) )
        {
            // Check whether this is a combination of m elements.
            if ( i.bitCount() == m )
            {
                q = q.add( BigInteger.ONE );
                Object[] combination = new Object[n];
                String empty = "-";
                    
                // For all n elements test whether element j is present.
                for ( int j = 0; j < n; ++j )
                {
                    // If present add it to the current combination in reverse order.  Reverse the
                    // order to take advantage of the fact that less significant bits change faster when
                    // incrementing or decrementing.
                    if ( i.testBit( j ) )
                        combination[(n - 1) - j] = elements[(n - 1) - j];
                    else
                        combination[(n - 1) - j] = empty;
                }

                // Process the current combination.
                listCombination( combination );
            }
        }
    }

    
    /** machine order for all m
     * @param elements The set of elements
     * @param n The total number of elements
     * @param t The total number of nonrepeating combinations
     */
    private static void machineOrder( Object[] elements, int n, BigInteger t )
    {
        // Iterate through all combinations of nonrepeating elements, incrementing by one.
        for ( BigInteger i = new BigInteger( "0" ); i.compareTo( t ) < 0; i = i.add( BigInteger.ONE ) )
        {
            Object[] combination = new Object[n];
            String empty = "-";
                    
            // For all n elements test whether element j is present.
            for ( int j = 0; j < n; ++j )
            {
                // If present add it to the current combination.
                if ( i.testBit( j ) )
                    combination[j] = elements[j];
                else
                    combination[j] = empty;
            }

            // List the current combination.
            listCombination( combination );
        }
    }


    /** machine order for a specific m
     * @param elements The set of elements
     * @param n The total number of elements
     * @param t The total number of nonrepeating combinations
     * @param m The number of elements to be combined
     */
    private static void machineOrder( Object[] elements, int n, BigInteger t, int m )
    {
        // total number of combinations with m elements
        BigInteger b = binomialCoefficient( n, m );    
        // current count of combinations with m elements
        BigInteger q = BigInteger.ZERO;

        // Iterate through all combinations of nonrepeating elements, incrementing by one.
        for ( BigInteger i = new BigInteger( "0" );
              i.compareTo( t ) < 0 && q.compareTo( b ) < 0;
              i = i.add( BigInteger.ONE ) )
        {
            if ( i.bitCount() == m )
            {
                q = q.add( BigInteger.ONE );
                Object[] combination = new Object[n];
                String empty = "-";
                    
                // For all n elements test whether element j is present.
                for ( int j = 0; j < n; ++j )
                {
                    // If present add it to the current combination.
                    if ( i.testBit( j ) )
                        combination[j] = elements[j];
                    else
                        combination[j] = empty;
                }

                // List the current combination.
                listCombination( combination );
            }
        }
    }


    /** Binomial coefficient used for optimization with large n
     * @param n The total number of elements
     * @param m The number of elements to be combined
     */
    private static BigInteger binomialCoefficient( int n, int m )
    {
        ArrayList<BigInteger> b = new ArrayList<>(n + 1);
        b.add( BigInteger.ONE );
        for ( int i = 1; i <= n; ++i )
        {
            b.add( BigInteger.ONE );
            for ( int j = i-1; j > 0; --j )
                b.set( j, b.get( j ).add( b.get( j - 1 ) ) );
        }

        return b.get( m );
    }


    /** Process the given combination. Here we just write it to the console.
     * @param combination The combination to process
     */
    private static void listCombination( Object[] combination )
    {
        int n = Array.getLength( combination );

        // For this combination list the elements that are present.
        for ( int j = 0; j < n; ++j )
        {
            System.out.print( combination[j].toString() );
            System.out.print( " " ); // element separator
        }
        System.out.println( "" ); // end of this combination
    }
}

